src/Development/Shake/Internal/Rules/File.hs

```diff
@@ -58,6 +58,13 @@ newtype FileQ = FileQ {fromFileQ :: FileName}
 data FileA = FileA {-# UNPACK #-} !ModTime {-# UNPACK #-} !FileSize FileHash
     deriving (Typeable,Eq)
 
+-- | Result of a File rule, may contain raw file information and wether the rule did run this build
+data FileR = FileR (Maybe FileA) Bool
```

Comment:
I'd have probably gone for a pair, although in many ways a data type is "cleaner". If going for a data type, I'd use record names, so `hasChanged` becomes a record, and also an annotation on the data type.

Reply:
Also make sure to say the Bool isn't serialised and gets recreated at False, doesn't participate in Eq/Hash etc.

Reply:
I chose the record for now. I can change that to a pair, if you prefer that.

Reply:
I remember now, why I chose a plain data type last time. Travis slaps me for not using the first record field ...